### PR TITLE
Say goodbye when terminating

### PR DESCRIPTION
**Pull request by joerg-krause**
Update the dependencies and say goodbye to the Spotify Connect client when librespot is terminate so it stops playback and removes the device from its list.

Closes: #114

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #![crate_name = "librespot"]
 
+#![feature(const_fn)]
```
Comment (plietar) [inline src/lib.rs]: That would break builds on stable

File: src/player.rs [hunk 2]
```diff
@@ -11,6 +13,18 @@ use metadata::{FileFormat, Track, TrackRef};
...
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::process::exit;
+
+use nix::sys::signal;
+
+static mut received_signal: AtomicUsize = AtomicUsize::new(0);
+
+extern fn sighandler_exit(_: libc::c_int) {
+    unsafe {
+        received_signal.store(15, Ordering::Relaxed);
```
Comment (plietar) [inline src/player.rs]: 15 ?

File: src/player.rs [hunk 3]
```diff
@@ -11,6 +13,18 @@ use metadata::{FileFormat, Track, TrackRef};
 use session::{Bitrate, Session};
 use util::{self, ReadSeek, SpotifyId, Subfile};
 pub use spirc::PlayStatus;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::process::exit;
+
+use nix::sys::signal;
+
+static mut received_signal: AtomicUsize = AtomicUsize::new(0);
```
Comment (plietar) [inline src/player.rs]: You can use `ATOMIC_USIZE_INIT`, which would remove the need for `const_fn`.
Also why a usize rather a bool ?

**Review by plietar — APPROVED** [review]
`nix` is really too low level for this. You could use something like https://github.com/tailhook/signal.
The player shouldn't have anything to do with this. I'm generally not comfortable having anything outside of main which has a global effect.
I would do all the signal stuff in main, just call `spirc.n otify_goobye()` and exit.

File: src/player.rs (hunk 3 above)
Comment (joerg-krause) [reply src/player.rs]: Agreed. I will replace it. I would like to add support for handling more signals, thats why an integer.

File: src/player.rs (hunk 2 above)
Comment (joerg-krause) [reply src/player.rs]: 15 is the portable number for SIGTERM.

**Comment by joerg-krause** [comment]
My first try was to run it in main, but I moved it to the player loop, because this was the only option to check for a signal.

I will play with signal and chan-signal and see if I am able to use them with my limited Rust knowledge :smile:

**Comment by joerg-krause** [comment]
@plietar I need some help!

I am trying to use [simple-signal](https://crates.io/crates/simple-signal):

``` rust
    let spirc = SpircManager::new(session.clone(), player);
    thread::spawn(move || spirc.run());
    Signals::set_handler(&[Signal::Int, Signal::Term],
        move |signals| {
            println!("Caught: {:?}", signals);
            spirc.send_goodbye();
            exit(0);
        }
    );
```

This fails with: `error[E0382]: capture of moved value: spirc`.

Unfortunately, my Rust skills are still limited, so can you give me a hint?

**Comment by joerg-krause** [comment]
Is it as easy as to clone spirc? Like this:

``` rust
    let spirc = SpircManager::new(session.clone(), player);
    let spirc_signal = spirc.clone();
    thread::spawn(move || spirc.run());
    Signals::set_handler(&[Signal::Int, Signal::Term],
        move |signals| {
            println!("Caught: {:?}", signals);
            spirc_signal.send_goodbye();
            exit(0);
        }
    );
```

Or should it be done in a different way?

**Comment by plietar** [comment]
Yeah that looks good. `SpircManager` is reference counted and can be cloned and shared between threads.

**Comment by joerg-krause** [comment]
Thanks! I've updated the commit...

**Pull request merged**